# **Waze App User Churn Prediction | Hypothesis Testing**

## Goal: 
* Conduct a two-sample hypothesis test (t-test) to analyze and determine whether there is a statistically significant difference in the average number of drives between iPhone users and Android users.

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

## **PACE: Plan**

### Research question: Is there a statistically significant difference in the mean amount of rides between iPhone users and Android users? 
    
#### "Do drivers who open the application using an iPhone have the same number of drives on average as drivers who use Android devices?"

### **Task 1. Imports and data loading**




In [5]:
import pandas as pd
import numpy as np
from scipy import stats
import seaborn as sns
import matplotlib.pyplot as plt

In [6]:
# Load dataset into dataframe
df = pd.read_csv('waze_dataset.csv')

In [7]:
df.head()

ID     label  sessions  drives  total_sessions  n_days_after_onboarding  \
0   0  retained       283     226      296.748273                     2276   
1   1  retained       133     107      326.896596                     1225   
2   2  retained       114      95      135.522926                     2651   
3   3  retained        49      40       67.589221                       15   
4   4  retained        84      68      168.247020                     1562   

   total_navigations_fav1  total_navigations_fav2  driven_km_drives  \
0                     208                       0       2628.845068   
1                      19                      64      13715.920550   
2                       0                       0       3059.148818   
3                     322                       7        913.591123   
4                     166                       5       3950.202008   

   duration_minutes_drives  activity_days  driving_days   device  
0              1985.775061             28            19  Android  
1              3160.472914             13            11   iPhone  
2              1610.735904             14             8  Android  
3               587.196542              7             3   iPhone  
4              1219.555924             27            18  Android

## **PACE: Analyze and Construct**

### **Task 2. Data exploration**

#### Using descriptive statistics to conduct exploratory data analysis (EDA):

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14999 entries, 0 to 14998
Data columns (total 13 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   ID                       14999 non-null  int64  
 1   label                    14299 non-null  object 
 2   sessions                 14999 non-null  int64  
 3   drives                   14999 non-null  int64  
 4   total_sessions           14999 non-null  float64
 5   n_days_after_onboarding  14999 non-null  int64  
 6   total_navigations_fav1   14999 non-null  int64  
 7   total_navigations_fav2   14999 non-null  int64  
 8   driven_km_drives         14999 non-null  float64
 9   duration_minutes_drives  14999 non-null  float64
 10  activity_days            14999 non-null  int64  
 11  driving_days             14999 non-null  int64  
 12  device                   14999 non-null  object 
dtypes: float64(3), int64(8), object(2)
memory usage: 1.5+ MB


##### Adding a new column `device_type` and converting the categorical variable `device` to numeric variable:

* `1` for an `iPhone`
* `2` for `Android`

In [9]:
# 1. Create `map_dictionary`
map_dictionary = {'Android': 2, 'iPhone':1}

# 2. Create new `device_type` column
df['device_type'] = df['device']

# 3. Map the new column to the dictionary
df['device_type'] = df['device_type'].map(map_dictionary)
df['device_type'].head()

0    2
1    1
2    2
3    1
4    2
Name: device_type, dtype: int64

##### Calculating the average number of drives between device type: 

In [10]:
df.groupby(['device_type'])['drives'].mean()

device_type
1    67.859078
2    66.231838
Name: drives, dtype: float64

> Based on the calculations, it appears that drivers who use an iPhone device to interact with the application have a higher number of drives on average. However, this difference might arise from random sampling, rather than being a true difference in the number of drives. To assess whether the difference is statistically significant, conducting a hypothesis test is needed.


### **Task 3. Hypothesis testing**

Steps for conducting a hypothesis test:


1.   State the null hypothesis and the alternative hypothesis
2.   Choose a signficance level
3.   Find the p-value
4.   Reject or fail to reject the null hypothesis

### Step 1: State the null hypothesis and the alternative hypothesis 
* $H_0$: There is no difference in average number of drives between drivers who use iPhone devices and drivers who use Androids. 
* $H_A$: There is a difference in average number of drives between drivers who use iPhone devices and drivers who use Androids.

### Step 2: Choose a significance level
* Significance level: 5%
> This means that there is a 5% probability of rejecting the null hypothesis, assuming it is true. In words, a significance level of 5% means that you are willing to accept a 5% chance that you are wrong when you reject the null hypothesis. 

### Step 3: Find the p-value
#### Hypothesis testing method: Welch's t-test using `stats.ttest_ind() function`
##### Assumptions when using the Welch's t-test:
1. *The two groups are independent of each other.*
> Since each row in the dataset represents one unique user, the Android users and the iPhone users are independent of each other. 
2. *The variances between two groups are not equal.*
> The default argument `equal_var` in `stats.ttest_ind()` is `True`, which assumes population variances are equal. This equal variance assumption might not hold in practice since there is no strong reason to assume that the two groups (iPhone users vs Android users) have the same variance. In this analysis, the argument `equal_var` is set to `False`. 

#### Performing the t-test: 

In [13]:
# 1. Isolate the `drives` column for iPhone users.
iPhone_users = df[df['device_type']==1]['drives']

# 2. Isolate the `drives` column for Android users.
Android_users = df[df['device_type']==2]['drives']

# 3. Perform the t-test
stats.ttest_ind(a=iPhone_users, b=Android_users, equal_var=False)

TtestResult(statistic=1.463523206885235, pvalue=0.143351972680206, df=11345.066049381952)

### Step 4: Reject or fail to reject the null hypothesis

> The p-value of 0.14335 is greater than the significance level of 0.05, which indicates the high probability of such extreme observations in the user dataset and that the null hypothesis cannot be rejected. 
>
>  ***Fail to reject the null hypothesis***. Therefore, there is not a statistically significant difference in the average number of drives between iPhone users and Android users.
> > **Note:** Statistical significance refers to the confidence that the observed result is not due to chance. 

## **PACE: Execute**

### Insights from the hypothesis test:  
* Drivers who use iPhone devices on average have a similar number of drives as those who use Androids.
  
* One potential next step is to explore what other factors influence the variation in the number of drives, and run additional hypothesis tests to learn more about user behavior. Further, since the user experience is the same, temporary changes in marketing or user interface for the Waze app may provide more data to investigate churn. 

## [Link to the Executive Summary: Two-Sample Hypothesis Test Results](https://docs.google.com/presentation/d/1hs6YRyOOht6XBvi8IGzxcaKh8cBVWGkM6GjI7w0vDIk/edit?usp=sharing)